### PR TITLE
Adds a shell script to help build the H3 toolchains

### PR DESCRIPTION
**Pull request by zwoop**

### CODE REVIEW

File: tools/build_h3_tools.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+echo "|   sudo yum -y install libxml2-devel c-ares-devel libevent-deve          |"
+echo "|   sudo yum -y install jansson-devel zlib-devel systemd-devel            |"
+echo "+-------------------------------------------------------------------------+"
+echo
+echo
+
+
+# OpenSSL needs special hackery ...
+echo "Building OpenSSL with QUIC support"
+git clone https://github.com/tatsuhiro-t/openssl openssl-quic
```
Comment (bryancall) [inline tools/build_h3_tools.sh:38]: You can do a if [ ! -d openssl-quic ]; then

That way you don't have to check it out every time.  I also changing the clone to g`it clone -b OpenSSL_1_1_1g-quic-draft-29 --depth 1 git@github.com:tatsuhiro-t/openssl.git` will be a lot faster for people on slow internet connections (e.g. me).

**Comment by maskit** [comment]
Maybe too much to ask but it would be great if the script also build curl with H3 support so you can run autest.

**Comment by zwoop** [comment]
> Maybe too much to ask but it would be great if the script also build curl with H3 support so you can run autest.

Can we add that later ?

**Comment by maskit** [comment]
Sure, I can add it later.

File: tools/build_h3_tools.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+
+# Finally, nghttp2 which has support for H3
+echo "Building nghttp2 ..."
+[ ! -d nghttp2 ] && git clone https://github.com/tatsuhiro-t/nghttp2.git
+cd nghttp2
+git checkout --track -b quic origin/quic
+autoreconf -if
+./configure --prefix=${BASE} PKG_CONFIG_PATH=${BASE}/lib/pkgconfig:${OPENSSL}/lib/pkgconfig CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}" LDFLAGS="${LDFLAGS}"
+gmake -j
```
Comment (maskit) [inline tools/build_h3_tools.sh:80]: Build failed because of too many jobs. I had to pass `-j $(nproc)`.

**Comment by maskit** [comment]
With the limit for the number of jobs, I confirmed that tools are built and works.

**Comment by maskit** [comment]
This builds curl, but I had to modify the code to make it speak draft-29 (it seems like the latest code is broken to me).
```
echo "Building curl ..."
[ ! -d curl ] && git clone https://github.com/curl/curl.git
cd curl
autoreconf -i
./configure --prefix=${BASE} --with-ssl=${OPENSSL} --with-nghttp2=${BASE} --with-nghttp3=${BASE} --with-ngtcp2=${BASE} CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}" LDFLAGS="${LDFLAGS}"
gmake -j
sudo gmake install
```

**Comment by zwoop** [comment]
I added the curl section, but when I try it, it hangs:

```
odin (15:21) 277/0 $ /opt/bin/curl --http3 https://www.google.com/
```

**Comment by maskit** [comment]
Hangs? Mine simply failed. Anyway, the build step is correct. It'll probably be fixed at some point.

**Review by maskit — APPROVED** [review]

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**